### ISSUE TITLE
Add Docstrings to each t_func

### ISSUE THREAD

**Issue by tpogden** [tpogden/maxwellbloch#25]

**Comment by amcdawes** [comment]
I'm happy to take a shot at this, but wonder if there is any progress on an alternative way to implement the t_funcs so that they don't need to be duplicated so much. One option would be to use a generator so the solver generated a unique t_func as needed. I am still getting my head around generators but will play around with this to see if it could help in this case.

**Comment by tommyogden** [comment]
@amcdawes Yes, this does need fixing soon. I raised #38 for that, with a comment about the best place to get around the problem[0], but if you think of another way that'd be great. :)

[0] The problem being that all the time-dependent args need to be passed in the same dict, and so need to be differentiated (e.g. with two pulses, we can't have that dict including {'width': 1, 'width': 2}).

**Comment by tommyogden** [comment]
Hi @amcdawes, I think your original idea of using a closure (nested function) is the way to go. We would replace e.g. `square_1`, `square_2` in the t_funcs module with something like:

```py
def bind_square(idx):

    def square(t, args):       
        on = args['on'+str(idx)]
        off = args['off'+str(idx)]
        ampl = args['ampl'+str(idx)]

        return ampl*(t >= on)*(t <= off)

    square.__name__ = 'square' + str(idx) # Not really needed?
    return square 
```

so that 

```py
square1 = bind_square(1)
```

generates us an indexed function
```py
square1(t=2.0, args={'ampl1':3.0, 'on1':1.0, 'off1':4.0})
Out: 3.0
```

The args are fairly straightfoward, in that we just need to change the dict keys to include an index. So If we have...

```py
field_0_func = 'square'
field_0_args = {'on':0.0, 'off':1.0, 'ampl':2.0}

field_1_func = 'square'
field_1_args = {'on':0.0, 'off':1.0, 'ampl':2.0}
```

we could put these into a list

```py
field_args_list = [field_0_args, field_1_args]
```

and then use the field list index to produce a single list of indexed args:

```py
args_dict = {}
for i, args_i in enumerate(field_args_list):
    for key, value in args_i.items():
        args_dict[key+str(i)] = value      
print(args_dict)
Out: {'ampl1': 2.0, 'on0': 0.0, 'ampl0': 2.0, 'off1': 1.0, 'off0': 1.0, 'on1': 0.0}
```

What do you think?

**Comment by tpogden** [comment]
Resolved by recent refactoring work (PRs #241–#248) or stale with no actionable body. Closing.
